app/controllers/projects/settings/project_custom_fields_controller.rb

```diff
@@ -0,0 +1,118 @@
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class Projects::Settings::ProjectCustomFieldsController < Projects::SettingsController
+  include OpTurbo::ComponentStream
+  include Projects::Settings::ProjectCustomFields::ComponentStreams
+
+  menu_item :settings_project_custom_fields
+
+  before_action :eager_load_project_custom_field_sections, only: %i[show toggle enable_all_of_section disable_all_of_section]
+  before_action :eager_load_project_custom_fields, only: %i[show toggle enable_all_of_section disable_all_of_section]
+  before_action :eager_load_project_custom_field_project_mappings,
+                only: %i[show toggle enable_all_of_section disable_all_of_section]
+
+  before_action :set_project_custom_field_section, only: %i[enable_all_of_section disable_all_of_section]
+
+  def show; end
+
+  def toggle
+    call = ProjectCustomFieldProjectMappings::ToggleService
+      .new(user: current_user)
+      .call(permitted_params.project_custom_field_project_mapping)
+
+    # we don't need to rerender a component as the toggle switch shows the correct state base on a successful response
+    if call.success?
+      render json: {}, status: :ok
+    else
+      render json: {}, status: :unprocessable_entity
+    end
+  end
+
+  def enable_all_of_section
+    call = bulk_edit_service.call(action: :enable)
+
+    if call.success?
+      eager_load_project_custom_field_project_mappings # reload mappings
+
+      update_sections_via_turbo_stream # update all sections in order not to mess with stimulus target references
+    else
+      # TODO: handle error
+    end
+
+    respond_with_turbo_streams
+  end
+
+  def disable_all_of_section
+    call = bulk_edit_service.call(action: :disable)
+
+    if call.success?
+      eager_load_project_custom_field_project_mappings # reload mappings
+
+      update_sections_via_turbo_stream # update all sections in order not to mess with stimulus target references
+    else
+      # TODO: handle error
+    end
+
+    respond_with_turbo_streams
+  end
+
+  private
+
+  def eager_load_project_custom_field_sections
+    @project_custom_field_sections = ProjectCustomFieldSection.all.to_a
```

Comment:
Is there a reason we need to explicitly call `to_a`? ActiveRecord seems to be loading a query only once.

Reply:
From my understanding it is required in order to force the data to be fetched from the database and to be loaded in memory. I'm doing this in order to avoid N+1 query in view components afterwards.

Reply:
I believe that will not prevent any N+1 queries, because when AR loads the results, it will not query the database each time the results are accessed.
However, we should definitely eager load the `custom_fields` association on the query, because it is referenced in the [project_custom_field_sections/show_component](https://github.com/opf/openproject/pull/14525/files#diff-ab7773fd1ebcdc5a97df0b4a20c6afe221b66a9be1c15d3ed51b0e2df5cfc0d5R40), which will indeed cause an N+1.

```suggestion
    @project_custom_field_sections = ProjectCustomFieldSection.includes(:custom_fields)
                                                         .order("custom_fields.position_in_custom_field_section ASC")
                                                         .all
```

Note: in order to not accidentally trigger the N+1 we should remove the `reorder(position_in_custom_field_section: :asc)` from [this line](https://github.com/opf/openproject/pull/14525/files#diff-ab7773fd1ebcdc5a97df0b4a20c6afe221b66a9be1c15d3ed51b0e2df5cfc0d5R40).